### PR TITLE
Add Free

### PR DESCRIPTION
**Pull request by hrb90**
Adds an instance for the `Free` group on a given set of generators. 

Does require a couple of new dependencies (`lists` and `foldable-traversable`) 

Not sure if there are any other instances that `Free` should have.

### CODE REVIEW

**Comment by hrb90** [comment]
Some Slack conversation exposed another construction we might consider for implementation of free groups:

`data FreeGrp a = GEmpty | GConsPlus a (FreeGrp a) | GConsMinus a (FreeGrp a)`

I'm not sure which would be better from an API design perspective and want to solicit feedback :)

File: src/Data/Group/Free.purs [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+-- Reduce a term of a free group to canonical form.
+canonical :: forall a. Eq a => List (Signed a) -> List (Signed a)
+canonical = foldr cancelOrPush Nil where
+  cancelOrPush x@(Positive x1) y@(Cons (Negative y1) tl) = if x1 == y1 then tl else x : y
+  cancelOrPush x@(Negative x1) y@(Cons (Positive y1) tl) = if x1 == y1 then tl else x : y
+  cancelOrPush x y = x : y
+
+-- The free group generated by elements of a, up to equality.
+newtype Free a = Free (List (Signed a))
```
Comment (morganthomas) [inline src/Data/Group/Free.purs]: Maybe this should be named FreeGroup, to distinguish from other free things, such as free monads. Currently free monads are also referred to as Free... https://github.com/purescript/purescript-free/blob/master/src/Control/Monad/Free.purs I'm not sure of the best approach.

**Comment by morganthomas** [comment]
Good idea! This looks right to me from a math perspective.

**Comment by morganthomas** [comment]
In my opinion the current approach is nicer than the one suggested on Slack. But is there a reason to prefer the Slack suggestion?

File: src/Data/Group/Free.purs (hunk 1 above)
Comment (hrb90) [reply src/Data/Group/Free.purs]: haskell's free group library calls this `Word`; I'd be okay with either `Free` or `FreeGroup`, but `Word` might be a little too mathematical and I'd rather make the connection with other free constructions explicit.

**Comment by hrb90** [comment]
Yes, I like the `Signed` approach better as well (and the [Haskell library](https://hackage.haskell.org/package/combinat-0.2.6.2/docs/Math-Combinat-FreeGroups.html) that exposes free groups does the same thing, for whatever that's worth.)

The upside of the `GConsPlus/GConsMinus` construction was that it might make the parallels to other free constructions even clearer, in that it "lifts the laws of the algebraic structure to data constructors" or something along those lines. I'm sorry if that's not too clear; I'm probably not a great advocate for something I don't really want to reimplement.

By the way, I went ahead and published this as its own library on Pursuit under the name `purescript-free-group`. (What can I say, I was excited to publish a library.) I still think it probably fits better as a part of `purescript-group`, and I'll deprecate the package if this PR gets accepted, but I would like to port over the changes I've made to this branch.

**Comment by morganthomas** [comment]
@hrb90 Let me know when(/if) you have something you feel is ready to be merged. Thank you! This is a very nice addition to the package imo.

**Comment by hrb90** [comment]
Updated the code, think it's ready to merge :smile:

**Review by morganthomas — APPROVED** [review]
@hrb90 This looks good to me! I'm going to merge this and rename Free to FreeGroup. Thanks again! Sorry for the delay.

**Comment by morganthomas** [comment]
Merged, changed names as described, bumped version, published to Pursuit!

**Pull request merged**